File: solution/java/week05-complexity/exercise2/src/test/java/command/BirthdayCommand.java
package command;

import java.time.LocalDate;
import java.time.temporal.ChronoUnit;
import java.util.Date;
import java.util.function.Supplier;

public class BirthdayCommand implements Command {
    private final LocalDate today;
    private final LocalDate birthday;

    public BirthdayCommand(Supplier<LocalDate> today, LocalDate birthday) {
        this.today = today.get();
        this.birthday = birthday;
    }

    @Override
    public String executeAndDisplayResult() {
        long daysBetween = ChronoUnit.DAYS.between(today, birthday);
        if (daysBetween < 0) {  // If birthday passed this year
            // Calculate days until next year's birthday
            daysBetween = ChronoUnit.DAYS.between(today, birthday.plusYears(1));
        }
        return "Still " + daysBetween + " days left!";
    }
}

File: solution/kotlin/week05-exercise2/src/main/kotlin/command/BirthdayCommand.kt
package command

import java.time.LocalDate
import java.time.temporal.ChronoUnit.DAYS
import java.util.function.Supplier

class BirthdayCommand(private val today: Supplier<LocalDate>, private val birthday: LocalDate) : Command {
    override fun executeAndDisplayResult(): String = "Still ${daysUntilBirthday()} days left!"

    private fun daysUntilBirthday(): Long =
        DAYS.between(today.get(), birthday)
            .let {
                if (it < 0) return DAYS.between(today.get(), birthday.plusYears(1))
                else return it
            }
}
